File: src/UnwindMC.Tests/CppEmitterTests.cs
using System.Collections.Generic;
using NUnit.Framework;
using UnwindMC.Analysis.Data;
using UnwindMC.Tests.Helpers;
using static UnwindMC.Tests.Helpers.AstHelper;

namespace UnwindMC.Tests
{
    [TestFixture, Ignore("Reimplement tests in F#")]
    public class CppEmitterTests
    {
        [Test]
        public void TestEmissionWithFunctionPointers()
        {
            var types = new Dictionary<string, Type>
            {
                { "arg0", new Type(true, 1) },
                { "arg1", new Type(true, 1) },
                { "var0", new Type(true, 1) },
                { "var1", new Type(true, 0) },
            };

            //var body = Scope(
            //    Assign(Var("var0"), Var("arg0")),
            //    While(Less(Var("var0"), Var("arg1")),
            //        Scope(
            //            Assign(Var("var1"), Dereference(Var("var0"))),
            //            IfThenElse(NotEqual(Var("var1"), Val(0)),
            //                Scope(Call(Var("var1"))),
            //                Scope()),
            //            Assign(Var("var0"), Add(Var("var0"), Val(1))))),
            //    Ret());

            //var source = CppEmitter.emit("foo", types, 2, body);
            //var expected =
            //    @"void foo(void (**arg0)(), void (**arg1)())
            //    {
            //      void (**var0)() = arg0;
            //      while (var0 < arg1)
            //      {
            //        void (*var1)() = *(var0);
            //        if (var1 != 0)
            //        {
            //          var1();
            //        }
            //        var0 = var0 + 1;
            //      }
            //      return;
            //    }
            //    ".StripIndent();
            //AssertSourceCodeEquals(expected, source);
        }

        [Test]
        public void TestEmissionFindMax()
        {
            var types = new Dictionary<string, Type>
            {
                { "arg0", new Type(false, 1) },
                { "arg1", new Type(false, 0) },
                { "var0", new Type(false, 0) },
                { "var1", new Type(false, 0) },
                { "var2", new Type(false, 1) },
                { "var3", new Type(false, 0) },
            };

            //var body = Scope(
            //    Assign(Var("var0"), Var("arg1")),
            //    Assign(Var("var1"), Val(int.MinValue)),
            //    IfThenElse(NotEqual(Var("var0"), Val(0)),
            //        Scope(
            //            Assign(Var("var2"), Var("arg0")),
            //            Assign(Var("var1"), Val(int.MinValue)),
            //            DoWhile(
            //                Scope(
            //                    Assign(Var("var3"), Dereference(Var("var2"))),
            //                    IfThenElse(Less(Var("var1"), Var("var3")),
            //                        Scope(Assign(Var("var1"), Var("var3"))),
            //                        Scope()),
            //                    Assign(Var("var2"), Add(Var("var2"), Val(1))),
            //                    Assign(Var("var0"), Subtract(Var("var0"), Val(1)))),
            //                NotEqual(Var("var0"), Val(0)))),
            //        Scope()),
            //    Ret(Var("var1")));

            //var source = CppEmitter.emit("findMax", types, 2, body);
            //// TODO: resolve return type/value
            //var expected =
            //    @"int findMax(int *arg0, int arg1)
            //    {
            //      int var0 = arg1;
            //      int var1 = -2147483648;
            //      if (var0 != 0)
            //      {
            //        int *var2 = arg0;
            //        var1 = -2147483648;
            //        do
            //        {
            //          int var3 = *(var2);
            //          if (var1 < var3)
            //          {
            //            var1 = var3;
            //          }
            //          var2 = var2 + 1;
            //          var0 = var0 - 1;
            //        } while (var0 != 0);
            //      }
            //      return var1;
            //    }
            //    ".StripIndent();
            //AssertSourceCodeEquals(expected, source);
        }

        private static void AssertSourceCodeEquals(string expected, string actual)
        {
            Assert.That(actual.Replace("\r\n", "\n"), Is.EqualTo(expected.Replace("\r\n", "\n")));
        }
    }
}

File: src/UnwindMC.Tests/CppEmitterTests.fs
﻿[<NUnit.Framework.TestFixture>]
module rec CppEmitterTests

open System
open System.Collections.Generic
open NUnit.Framework
open Ast
open Type
open MiscExtensions

[<Test>]
let testEmissionWithFunctionPointers (): unit =
    let body =
        [|
            Assignment (Var "var0", VarRef (Var "arg0"))
            While
                (
                    Binary (Operator.Less, VarRef (Var "var0"), VarRef (Var "arg1")),
                    [|
                        Assignment (Var "var1", Dereference (VarRef (Var "var0")))
                        IfThenElse
                            (
                                Binary (Operator.NotEqual, VarRef (Var "var1"), Expression.Value 0),
                                [| FunctionCall (VarRef (Var "var1")) |],
                                [||]
                            )
                        Assignment (Var "var0", Binary (Operator.Add, VarRef (Var "var0"), Expression.Value 1))
                    |]
                )
            Statement.Return None
        |]

    let source =
        CppEmitter.emit {
            name = "foo"
            parameters =
                [|
                    { name = "arg0"; type_ = { isFunction = true; indirectionLevel = 1; size = 4 } }
                    { name = "arg1"; type_ = { isFunction = true; indirectionLevel = 1; size = 4 } }
                |]
            locals = [||]
            variables =
                [|
                    { name = "var0"; type_ = { isFunction = true; indirectionLevel = 1; size = 4 } }
                    { name = "var1"; type_ = { isFunction = true; indirectionLevel = 0; size = 4 } }
                |]
            body = body
        }
    let expected =
        """void foo(void (**arg0)(), void (**arg1)())
        {
          void (**var0)();
          void (*var1)();
        
          var0 = arg0;
          while (var0 < arg1)
          {
            var1 = *(var0);
            if (var1 != 0)
            {
              var1();
            }
            var0 = var0 + 1;
          }
          return;
        }
        """.StripIndent()
    assertSourceCodeEquals expected source

[<Test>]
let testEmissionFindMax (): unit =
    let body =
        [|
            Assignment (Var "var0", VarRef (Var "arg1"))
            Assignment (Var "var1", Expression.Value Int32.MinValue)
            IfThenElse
                (
                    Binary (Operator.NotEqual, VarRef (Var "var0"), Expression.Value 0),
                    [|
                        Assignment (Var "var2", VarRef (Var "arg0"))
                        Assignment (Var "var1", Expression.Value Int32.MinValue)
                        DoWhile
                            (
                                [|
                                    Assignment (Var "var3", Dereference (VarRef (Var "var2")))
                                    IfThenElse
                                        (
                                            Binary (Operator.Less, VarRef (Var "var1"), VarRef (Var "var3")),
                                            [| Assignment (Var "var1", VarRef (Var "var3")) |],
                                            [||]
                                        )
                                    Assignment (Var "var2", Binary (Operator.Add, VarRef (Var "var2"), Expression.Value 1))
                                    Assignment (Var "var0", Binary (Operator.Subtract, VarRef (Var "var0"), Expression.Value 1))
                                |],
                                Binary (Operator.NotEqual, VarRef (Var "var0"), Expression.Value 0)
                            )
                    |],
                    [||]
                )
            Statement.Return (Some (Var "var1"))
        |]

    let source =
        CppEmitter.emit {
            name = "findMax"
            parameters =
                [|
                    { name = "arg0"; type_ = { isFunction = false; indirectionLevel = 1; size = 4 } }
                    { name = "arg1"; type_ = { isFunction = false; indirectionLevel = 0; size = 4 } }
                |]
            locals = [||]
            variables =
                [|
                    { name = "var0"; type_ =  { isFunction = false; indirectionLevel = 0; size = 4 } }
                    { name = "var1"; type_ =  { isFunction = false; indirectionLevel = 0; size = 4 } }
                    { name = "var2"; type_ =  { isFunction = false; indirectionLevel = 1; size = 4 } }
                    { name = "var3"; type_ =  { isFunction = false; indirectionLevel = 0; size = 4 } }
                |]
            body = body
        }
    let expected =
        """int findMax(int *arg0, int arg1)
        {
          int var0;
          int var1;
          int *var2;
          int var3;
        
          var0 = arg1;
          var1 = -2147483648;
          if (var0 != 0)
          {
            var2 = arg0;
            var1 = -2147483648;
            do
            {
              var3 = *(var2);
              if (var1 < var3)
              {
                var1 = var3;
              }
              var2 = var2 + 1;
              var0 = var0 - 1;
            } while (var0 != 0);
          }
          return var1;
        }
        """.StripIndent()
    assertSourceCodeEquals expected source

let assertSourceCodeEquals (expected: string) (actual: string): unit =
    Assert.That(actual.Replace("\r\n", "\n"), Is.EqualTo(expected.Replace("\r\n", "\n")))
